### PR TITLE
Add client and server lookup throttling

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation

Sometimes, broker requires a way to throttle newly incoming traffic and requires to restrict number of incoming concurrent lookup requests. It is also useful to throttle at client in order to avoid large number of concurrent lookup-requests while creating producers/consumers.

### Modifications

- Add broker side throttling to serve max allowed concurrent lookup request
- Add client side lookup throttling
- Update value of maxConcurrentLookupRequest dynamically at broker
- AdminApi command to update maxConcurrentLookupRequest

### Result

Broker can have capability to throttle newly incoming traffic by restricting number of concurrent lookup request.

### CODE REVIEW

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/lookup/DestinationLookup.java [hunk 1]
```diff
@@ -31,26 +38,30 @@
...
-import static com.google.common.base.Preconditions.checkNotNull;
+import com.yahoo.pulsar.common.util.ObjectMapperFactory;
 
 import io.netty.buffer.ByteBuf;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiResponse;
+import io.swagger.annotations.ApiResponses;
 
 @Path("/v2/destination/")
 @NoSwaggerDocumentation
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/lookup/DestinationLookup.java]: added http-throttling change as well.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java [hunk 2]
```diff
@@ -820,6 +843,36 @@ public String generateUniqueProducerName() {
...
+        int pendingLookupRequest = pulsar.getConfiguration().getMaxConcurrentLookupRequest();
+        try {
+            Optional<Map<String, String>> data = dynamicConfigurationCache.get(LOOKUP_THROTTLING_PATH);
+            if (data.isPresent() && data.get().get(THROTTLING_LOOKUP_REQUEST_KEY) != null) {
+                pendingLookupRequest = Integer.parseInt(data.get().get(THROTTLING_LOOKUP_REQUEST_KEY));
+            }
+        } catch (Exception e) {
+            log.warn("Got exception when reading ZooKeeper path [{}]:", LOOKUP_THROTTLING_PATH, e);
+        }
+        return pendingLookupRequest;
```
Comment (rdhabalia) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/BrokerService.java]: added debug log.

File: pulsar-broker/src/main/java/com/yahoo/pulsar/broker/lookup/DestinationLookup.java [hunk 3]
```diff
@@ -98,30 +115,54 @@ public void lookupDestinationAsync(@PathParam("property") String property, @Path
...
 
     }
 
 
+    @PUT
+    @Path("permits/{permits}")
+    @ApiOperation(value = "Update allowed concurrent lookup permits. This operation requires Pulsar super-user privileges.")
+    @ApiResponses(value = { @ApiResponse(code = 204, message = "Lookup permits updated successfully"),
+            @ApiResponse(code = 403, message = "You don't have admin permission to create the cluster") })
+    public void createCluster(@PathParam("permits") int permits) {
```
Comment (merlimat) [inline pulsar-broker/src/main/java/com/yahoo/pulsar/broker/lookup/DestinationLookup.java]: Method should not be called `createCluster()`

**Comment by rdhabalia** [comment]
> The implementation for client side and broker side throttling is very different and ideally it would be better to break this down into 2 separate commits.

Agree. I will create separate commit with different PR to address it.

> While I agree that having variables dynamically tunable at runtime (without restarting the broker) is a great thing, we shouldn't be doing that on a case by case, adding the REST handlers (that would not be possible to remove later.. ). We should address the dynamic configuration in a separate PR, using a more general approach

Agree. Does it make sense to create a separate resource (REST controller) ```Configuration``` which allows to add/update/delete configuration: param1: zk_path,  body. json_value

**Comment by merlimat** [comment]
> Agree. Does it make sense to create a separate resource (REST controller) Configuration which allows to add/update/delete configuration: param1: zk_path, body. json_value

We need to think well how to expose the whole thing in consistent way. You have the values from the config file + the values in ZK. There should be a clear way to understand what is the actual configuration that is:
 * Supposed to be applied to all brokers
 * It's *actually* being applied in a given broker... 

Other problem is how to clearly identify which values are dynamically tunable (eg: zk servers string will not be dynamic...)

**Comment by rdhabalia** [comment]
@merlimat @saandrews this PR was having 3 changes
- server throttling
- client throttling
- updating throttling-configuration using REST-Api/AdminApi

I have created separate PR for #181(server) and #182(client).
As @merlimat mentioned: For dynamic-configuration update: we can separate PR to address generic configuration update.

**Comment by merlimat** [comment]
Closing in favor of #181

**Pull request closed without merging**